ArchiSteamFarm.Tests/Bot.cs

```diff
@@ -0,0 +1,126 @@
+// limitations under the License.
+
+using System.Collections.Generic;
+using ArchiSteamFarm.Json;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using static ArchiSteamFarm.Bot;
+
+namespace ArchiSteamFarm.Tests {
+	[TestClass]
+	public sealed class Bot {
+		[TestMethod]
+		public void NotAllCardsPresent() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset>? itemsToSend = GetItemsForFullBadge(items, 3);
+			Assert.IsNotNull(itemsToSend);
+			Assert.IsTrue(itemsToSend.Count == 0);
+		}
+
+		[TestMethod]
+		public void OneSet() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset>? itemsToSend = GetItemsForFullBadge(items, 2);
+			Assert.IsNotNull(itemsToSend);
+			Assert.IsTrue(itemsToSend.Count == items.Count);
+		}
+
+		[TestMethod]
+		public void MultipleSets() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1),
+				CreateCard(1),
+				CreateCard(2),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset>? itemsToSend = GetItemsForFullBadge(items, 2);
+			Assert.IsNotNull(itemsToSend);
+			Assert.IsTrue(itemsToSend.Count == items.Count);
+		}
+
+		[TestMethod]
+		public void MultipleSetsDifferentAmount() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, 2),
+				CreateCard(2),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset>? itemsToSend = GetItemsForFullBadge(items, 2);
+			Assert.IsNotNull(itemsToSend);
+			Assert.IsTrue(itemsToSend.Count == items.Count);
+		}
+
+		[TestMethod]
+		public void MoreCardsThanNeeded() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1),
+				CreateCard(1),
+				CreateCard(2),
+				CreateCard(3),
+			};
+
+			HashSet<Steam.Asset>? itemsToSend = GetItemsForFullBadge(items, 3);
+			Assert.IsNotNull(itemsToSend);
+			Assert.IsTrue(itemsToSend.Count == 3);
+		}
+
+		[TestMethod]
+		public void TooHighAmount() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, 2),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset>? itemsToSend = GetItemsForFullBadge(items, 2);
+			Assert.IsNotNull(itemsToSend);
+			Assert.IsTrue(itemsToSend.Count == 0);
+		}
+
+		[TestMethod]
+		public void PartiallyHighAmount() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, 5),
+				CreateCard(1, 4),
+				CreateCard(1),
+				CreateCard(2, 2),
+				CreateCard(2, 2),
+				CreateCard(2, 2),
+				CreateCard(2, 2),
+			};
+
+			HashSet<Steam.Asset>? itemsToSend = GetItemsForFullBadge(items, 2);
+			Assert.IsNotNull(itemsToSend);
+			Assert.IsTrue(itemsToSend.Count == 5);
+		}
+
+		private static Steam.Asset CreateCard(ulong classID, uint amount = 1) => new Steam.Asset(Steam.Asset.SteamAppID, Steam.Asset.SteamCommunityContextID, classID, amount, realAppID: 42, type: Steam.Asset.EType.TradingCard, rarity: Steam.Asset.ERarity.Common);
```

Comment:
You're not testing here at least two completely valid use cases:
- Cards from different games
- Foils vs Non-foils mixed

And you should also test with different `rarity` to ensure this code continues to work if Valve by any chance introduce cards with different rarities.

Reply:
The method tested will never get cards of different type (foil vs non-foil) or from different games.

Reply:
Then make it throw if something like that happens and add unit test proving it does.

Reply:
Even though I find it easier to just properly group by and make it a valid feature